apps/api/src/recurring-donation/recurring-donation.service.ts

```diff
@@ -39,15 +144,60 @@ export class RecurringDonationService {
         extCustomerId: updateRecurringDonationDto.extCustomerId,
         amount: updateRecurringDonationDto.amount,
         currency: updateRecurringDonationDto.currency,
+        vaultId: updateRecurringDonationDto.vaultId,
       },
     })
     if (!result) throw new NotFoundException('Not found')
     return result
   }
 
+  async cancel(id: string): Promise<RecurringDonation | null> {
+    return await this.updateStatus(id, RecurringDonationStatus.canceled)
+  }
+
   async remove(id: string): Promise<RecurringDonation | null> {
     const result = await this.prisma.recurringDonation.delete({ where: { id: id } })
     if (!result) throw new NotFoundException('Not found')
     return result
   }
+
+  async findSubscriptionByExtId(extSubscriptionId: string): Promise<RecurringDonation | null> {
+    const result = await this.prisma.recurringDonation.findFirst({
+      where: { extSubscriptionId },
+      select: RecurringDonationService.defaultSelectFields,
+    })
+    return result
+  }
+
+  async updateStatus(
+    id: string,
+    status: RecurringDonationStatus,
+  ): Promise<RecurringDonation | null> {
+    const result = await this.prisma.recurringDonation.update({
+      where: { id: id },
+      data: {
+        status: status,
+      },
+    })
+    if (!result) throw new NotFoundException('Not found')
+    return result
+  }
+
+  async cancelSubscription(subscriptionId: string) {
+    Logger.log(`Canceling subscription with api request to cancel: ${subscriptionId}`)
+    return this.stripeClient.subscriptions.cancel(subscriptionId).then((result) => {
+      if (result.status !== 'canceled') {
+        Logger.log(`Subscription cancel attempt failed with status of ${result.id}: ${result.status}`)
+        return
+      }
+
+      // the webhook will handle this as well.
+      // but we cancel it here, in case the webhook is slow.
+      this.findSubscriptionByExtId(result.id).then((rd) => {
+        if (rd) {
+          return this.cancel(rd.id)
+        }
+      })
+    })
+  }
```

Comment:
Let's use `async/await` syntnax instead of `then()` promise callbacks

Reply:
OK